# LISTA 11 - PANDAS

## <ins>Ejercicios Obligatorios </ins>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from pandas import Series, DataFrame

## EJERCICIO 1

Crea un dataframe con 15 filas y 2 columnas. La primera columna se llamará x, la segunda y. Cada entrada será un número aleatorio (puedes usar random.normal() de la librería numpy). Guarda el dataframe en una variable llamada points.

In [26]:
import pandas as pd
import numpy as np
data = np.random.normal(size=(15, 2)) # Crea una sample de datos aleatorios de 15 filas en 2 columnas
points = pd.DataFrame(data, columns=['x', 'y']) # Crea un dataframe con la matrix de 15x2 y le pone de cabecera la lista de columns
print(points)

           x         y
0   0.209339  1.208477
1   0.730889 -0.319757
2   1.060358  0.221681
3  -0.029359 -0.584837
4  -2.842888 -0.951093
5   0.042998  1.442983
6  -0.563719 -1.112271
7  -0.287107  0.066881
8  -3.005483 -2.648041
9  -1.208865  0.354483
10  2.156314 -0.205128
11  1.089402  0.871555
12  0.778905 -0.327643
13 -0.412882 -0.215737
14  0.980865  1.285192


## EJERCICIO 2

Del dataframe points, muestra las filas cuyo valor en la columna x sea positivo.

In [31]:
print(points[points['x']>0]) # points['x']>0 genera una columna de booleanos que si se usa de indice indica si imprimir o no la fila

           x         y
0   0.209339  1.208477
1   0.730889 -0.319757
2   1.060358  0.221681
5   0.042998  1.442983
10  2.156314 -0.205128
11  1.089402  0.871555
12  0.778905 -0.327643
14  0.980865  1.285192


## EJERCICIO 3

In [32]:
print(points[points['y']<0])

           x         y
1   0.730889 -0.319757
3  -0.029359 -0.584837
4  -2.842888 -0.951093
6  -0.563719 -1.112271
8  -3.005483 -2.648041
10  2.156314 -0.205128
12  0.778905 -0.327643
13 -0.412882 -0.215737


Del dataframe points, muestra las filas cuyo valor en la columna y sea negativo.

## EJERCICIO 4

In [34]:
print(points[(points['x'] > 0) & (points['y'] > 0)]) # Hace un AND entre las columnas booleanas de x e y para sacar las que ambas sean positivas

           x         y
0   0.209339  1.208477
2   1.060358  0.221681
5   0.042998  1.442983
11  1.089402  0.871555
14  0.980865  1.285192


Del dataframe points, muestra las observaciones cuyos puntos (x, y) pertenezcan al primer cuadrante. 

## EJERCICIO 5

In [38]:
for i,row in points.iterrows():
    sino = "si" if (row['x'] > 0) and (row['y'] > 0) else "no"
    print(f"{i} El punto ({row['x']:.4f},{row['y']:.4f}) {sino} pertenece al primer cuadrante") # ..4f limita a 4 decimales y redondea

0 El punto (0.2093,1.2085) si pertenece al primer cuadrante
1 El punto (0.7309,-0.3198) no pertenece al primer cuadrante
2 El punto (1.0604,0.2217) si pertenece al primer cuadrante
3 El punto (-0.0294,-0.5848) no pertenece al primer cuadrante
4 El punto (-2.8429,-0.9511) no pertenece al primer cuadrante
5 El punto (0.0430,1.4430) si pertenece al primer cuadrante
6 El punto (-0.5637,-1.1123) no pertenece al primer cuadrante
7 El punto (-0.2871,0.0669) no pertenece al primer cuadrante
8 El punto (-3.0055,-2.6480) no pertenece al primer cuadrante
9 El punto (-1.2089,0.3545) no pertenece al primer cuadrante
10 El punto (2.1563,-0.2051) no pertenece al primer cuadrante
11 El punto (1.0894,0.8716) si pertenece al primer cuadrante
12 El punto (0.7789,-0.3276) no pertenece al primer cuadrante
13 El punto (-0.4129,-0.2157) no pertenece al primer cuadrante
14 El punto (0.9809,1.2852) si pertenece al primer cuadrante


Del dataframe points, muestra las observaciones de la forma: “El punto ({x}, {y}) {no/sí} pertenece al primer cuadrante”.

## EJERCICIO 6

In [48]:
header = ['word']
word = ["euro", "diez", "algas", "broma", "cicuta", "fatiga", "nachos", "jadeos", "hazañas", "boutique"]
words = pd.DataFrame(data=word,columns=header)
words['length'] = words['word'].apply(len) # Usa la función por cada item de cada fila para generar la columna

def countVocals(word):
    n = 0
    for letter in word:
        if letter.lower() in 'aeiouáéíóú':
            n += 1
    return n
    
words['vowels'] = words['word'].apply(countVocals)
words['consonants'] = words['length'] - words['vowels']
print(words)

       word  length  vowels  consonants
0      euro       4       3           1
1      diez       4       2           2
2     algas       5       2           3
3     broma       5       2           3
4    cicuta       6       3           3
5    fatiga       6       3           3
6    nachos       6       2           4
7    jadeos       6       3           3
8   hazañas       7       3           4
9  boutique       8       5           3


Crea un dataframe con 10 filas y 4 columnas. La primera columna se llamará word; la segunda, length; la tercera, vowels; y la última, consonants. La columna words contendrá las siguientes 10 palabras: “euro”, “diez” “algas”, “broma”, “cicuta”, “fatiga”, “nachos”, “jadeos”, “hazañas”, “boutique”. Las columnas length, vowels y consonants contendrán, respectivamente, la longitud, el total de vocales y el total de consonantes (añade las columnas calculando su valor por código). Guarda el dataframe en la variable words.

## EJERCICIO 7

In [49]:
for i,row in words.iterrows():
    print(f"{i} La palabra {row['word']} tiene {row['length']} letras, de las cuales {row['vowels']} son vocales y {row['consonants']} son consonantes.")

0 La palabra euro tiene 4 letras, de las cuales 3 son vocales y 1 son consonantes.
1 La palabra diez tiene 4 letras, de las cuales 2 son vocales y 2 son consonantes.
2 La palabra algas tiene 5 letras, de las cuales 2 son vocales y 3 son consonantes.
3 La palabra broma tiene 5 letras, de las cuales 2 son vocales y 3 son consonantes.
4 La palabra cicuta tiene 6 letras, de las cuales 3 son vocales y 3 son consonantes.
5 La palabra fatiga tiene 6 letras, de las cuales 3 son vocales y 3 son consonantes.
6 La palabra nachos tiene 6 letras, de las cuales 2 son vocales y 4 son consonantes.
7 La palabra jadeos tiene 6 letras, de las cuales 3 son vocales y 3 son consonantes.
8 La palabra hazañas tiene 7 letras, de las cuales 3 son vocales y 4 son consonantes.
9 La palabra boutique tiene 8 letras, de las cuales 5 son vocales y 3 son consonantes.


Muestra las 10 observaciones de words con el formato “La palabra {word} tiene {length} letras, de las cuales {vowels} son vocales y {consonants} consonantes”.

## EJERCICIO 8

In [50]:
for i,row in words.iterrows():
    if row['vowels'] == row['consonants']:
        print(f"La palabra {row['word']} tiene la misma cantidad de vocales y consonantes.")

La palabra diez tiene la misma cantidad de vocales y consonantes.
La palabra cicuta tiene la misma cantidad de vocales y consonantes.
La palabra fatiga tiene la misma cantidad de vocales y consonantes.
La palabra jadeos tiene la misma cantidad de vocales y consonantes.


Muestra aquellas observaciones de words que tienen el mismo número de vocales y consonantes.

## EJERCICIO 9

In [52]:
fotocasa_df = pd.read_csv("https://drive.google.com/uc?export=view&id=1DwvALe87aeDzTk7s6xGPDd6itJ65s4YE", index_col = 0)
fotocasa_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 617 entries, 0 to 616
Data columns (total 59 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Precio               598 non-null    object
 1   Título               616 non-null    object
 2   Descripción          588 non-null    object
 3   Fotos                582 non-null    object
 4   Dirección            597 non-null    object
 5   Inmobiliaria         617 non-null    bool  
 6   Anunciante           598 non-null    object
 7   Habitaciones         551 non-null    object
 8   Baños                592 non-null    object
 9   Superficie           586 non-null    object
 10  Tipo de inmueble     597 non-null    object
 11  Ascensor             494 non-null    object
 12  Amueblado            469 non-null    object
 13  Depósito             136 non-null    object
 14  Gastos de comunidad  390 non-null    object
 15  Consumo energía      598 non-null    object
 16  Emisiones    

El siguiente DataFrame contiene información de los pisos en alquiler en Murcia a fecha de 31/12/2020:
```
fotocasa_df = pd.read_csv("https://drive.google.com/uc?export=view&id=1DwvALe87aeDzTk7s6xGPDd6itJ65s4YE", index_col = 0)
```
Muestra el número de filas, columnas y valores que contiene el DataFrame cargado:

## EJERCICIO 10

In [53]:
fotocasa_df.head()

Precio                                     Título  \
0  A consultar                 Piso de alquiler en Centro   
1   500 € /mes          Apartamento de alquiler en Centro   
2   900 € /mes  Piso de alquiler en Doctor Marañón Centro   
3   750 € /mes                 Piso de alquiler en Centro   
4   800 € /mes                 Piso de alquiler en Centro   

                                         Descripción     Fotos  \
0  TRABAJADORES Y ESTUDIANTES ATENCION!!!! VIVIEN...  18 Fotos   
1  Maravilloso apartamento para entrar a vivir en...  21 Fotos   
2  Se alquila estupenda vivienda recién reformada...  18 Fotos   
3  Te presentamos este fabuloso apartamento situa...  30 Fotos   
4  Fantástica vivienda en alquiler ubicada muy pr...  28 Fotos   

               Dirección  Inmobiliaria                    Anunciante  \
0                 Centro          True      DPTO. COMERCIAL PISOSMUR   
1                 Centro          True                BREXTOM MURCIA   
2  Doctor Marañón Centro          True  ABAD SERVICIOS INMOBILIARIOS   
3                 Centro          True  ABAD SERVICIOS INMOBILIARIOS   
4                 Centro          True                BREXTOM MURCIA   

                        Habitaciones                              Baños  \
0  <span><span>4</span> habs.</span>  <span><span>2</span> baños</span>   
1   <span><span>1</span> hab.</span>   <span><span>1</span> baño</span>   
2  <span><span>4</span> habs.</span>  <span><span>2</span> baños</span>   
3  <span><span>2</span> habs.</span>   <span><span>1</span> baño</span>   
4  <span><span>3</span> habs.</span>  <span><span>2</span> baños</span>   

                         Superficie  ... Serv. portería Internet  \
0  <span><span>110</span> m²</span>  ...            NaN      NaN   
1   <span><span>46</span> m²</span>  ...            NaN      NaN   
2  <span><span>169</span> m²</span>  ...            NaN      NaN   
3   <span><span>77</span> m²</span>  ...            NaN      NaN   
4  <span><span>104</span> m²</span>  ...            NaN      NaN   

  Suite - con baño Patio Domótica Pista de Tenis Alarma Mascotas  \
0              NaN   NaN      NaN            NaN    NaN      NaN   
1              NaN   NaN      NaN            NaN    NaN      NaN   
2              NaN   NaN      NaN            NaN    NaN      NaN   
3              NaN   NaN      NaN            NaN    NaN      NaN   
4              NaN   NaN      NaN            NaN    NaN      NaN   

  Energía Solar Gimnasio  
0           NaN      NaN  
1           NaN      NaN  
2           NaN      NaN  
3           NaN      NaN  
4           NaN      NaN  

[5 rows x 59 columns]

Recupera los cinco primeros pisos (filas)

## EJERCICIO 11

In [54]:
fotocasa_df.tail()

Precio                                             Título Descripción  \
612    NaN  Alquiler piso en barrio dEl Carmen en El Carme...         NaN   
613    NaN  Alquiler apartamento en plaza san nicolas en M...         NaN   
614    NaN  Alquiler apartamento exclusivo apto plaza flor...         NaN   
615    NaN  Alquiler piso alquiler de local entresuelo en ...         NaN   
616    NaN  Alquiler piso en alquiler en edificio alba, sa...         NaN   

    Fotos Dirección  Inmobiliaria Anunciante Habitaciones Baños Superficie  \
612   NaN       NaN          True        NaN          NaN   NaN        NaN   
613   NaN       NaN          True        NaN          NaN   NaN        NaN   
614   NaN       NaN          True        NaN          NaN   NaN        NaN   
615   NaN       NaN          True        NaN          NaN   NaN        NaN   
616   NaN       NaN          True        NaN          NaN   NaN        NaN   

     ... Serv. portería Internet Suite - con baño Patio Domótica  \
612  ...            NaN      NaN              NaN   NaN      NaN   
613  ...            NaN      NaN              NaN   NaN      NaN   
614  ...            NaN      NaN              NaN   NaN      NaN   
615  ...            NaN      NaN              NaN   NaN      NaN   
616  ...            NaN      NaN              NaN   NaN      NaN   

    Pista de Tenis Alarma Mascotas Energía Solar Gimnasio  
612            NaN    NaN      NaN           NaN      NaN  
613            NaN    NaN      NaN           NaN      NaN  
614            NaN    NaN      NaN           NaN      NaN  
615            NaN    NaN      NaN           NaN      NaN  
616            NaN    NaN      NaN           NaN      NaN  

[5 rows x 59 columns]

Recupera los cinco últimos pisos:

Muestra los campos Inmobiliaria y Tipo de inmueble de los tres últimos pisos:

## EJERCICIO 12

In [55]:
fotocasa1_df = fotocasa_df.drop(labels = ["Domótica", "Pista de Tenis","Alarma","Mascotas","Energía Solar","Gimnasio"], axis = 1)
fotocasa1_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 617 entries, 0 to 616
Data columns (total 53 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Precio               598 non-null    object
 1   Título               616 non-null    object
 2   Descripción          588 non-null    object
 3   Fotos                582 non-null    object
 4   Dirección            597 non-null    object
 5   Inmobiliaria         617 non-null    bool  
 6   Anunciante           598 non-null    object
 7   Habitaciones         551 non-null    object
 8   Baños                592 non-null    object
 9   Superficie           586 non-null    object
 10  Tipo de inmueble     597 non-null    object
 11  Ascensor             494 non-null    object
 12  Amueblado            469 non-null    object
 13  Depósito             136 non-null    object
 14  Gastos de comunidad  390 non-null    object
 15  Consumo energía      598 non-null    object
 16  Emisiones    

Las columnas Domótica, Pista de Tenis, Alarma, Mascotas, Energía Solar y Gimnasio no contienen apenas valores definidos. Elimínalas del DataFrame y almacena el resultado en fotocasa1_df:

## EJERCICIO 13

In [56]:
fotocasa2_df = fotocasa1_df.rename(columns={'Tipo de inmueble': 'Tipo'})
print(fotocasa2_df['Tipo'].head())

0           Piso
1    Apartamento
2           Piso
3           Piso
4           Piso
Name: Tipo, dtype: object


Renombra a continuación la columna Tipo de inmueble de fotocasa1_df por Tipo. Asigna el resultado a la variable fotocasa2_df:

## EJERCICIO 14

In [63]:
fotocasa4_df = fotocasa2_df[fotocasa2_df['Precio'].notnull() & (fotocasa2_df['Precio'] != "A consultar")]
print(fotocasa2_df['Precio'])
print("\n\n")
fotocasa4_df['Precio']

0      A consultar
1       500 € /mes
2       900 € /mes
3       750 € /mes
4       800 € /mes
          ...     
612            NaN
613            NaN
614            NaN
615            NaN
616            NaN
Name: Precio, Length: 617, dtype: object





1      500 € /mes
2      900 € /mes
3      750 € /mes
4      800 € /mes
5      650 € /mes
          ...    
589    600 € /mes
590    800 € /mes
591    800 € /mes
593    800 € /mes
596    850 € /mes
Name: Precio, Length: 588, dtype: object

Elimina del DataFrame aquellos pisos que no tienen Precio o que su Precio sea "A consultar". Asigna el resultado a la variable fotocasa4_df:

PISTA: para seleccionar los alquileres sin precio utiliza el método isnull(). Utiliza notnull() para seleccionar los alquileres con precio..

## EJERCICIO 15

In [72]:
convert = ["Tipo","Orientación","Estado","Parking"]
fotocasa5_df = fotocasa4_df.copy()
fotocasa5_df[convert] = fotocasa4_df[convert].astype("category")
fotocasa5_df[convert].dtypes

Tipo           category
Orientación    category
Estado         category
Parking        category
dtype: object

Utiliza la función astype para convertir las columnas Tipo, Orientación, Estado y Parking a tipo categórico. Asigna los resultados a la variable fotocasa5_df y comprueba mediante el parámetro dtype que el tipo asignado sea correcto:

## EJERCICIO 16

Cargar el fichero situado en https://raw.githubusercontent.com/ainaramu-icjardin/big_data/main/ciudades_ejemplo.csv
como un dataframe Pandas, y guardarlo a continuación en formato excel.

## EJERCICIO 17

Cargar el fichero "https://raw.githubusercontent.com/ainaramu-icjardin/big_data/main/parocomunidades.csv" que está con codificación (encoding) "latin1" y dejarlo en un dataframe df_paro.

Mostrar solo las filas de `df_paro` que corresponden al Periodo 2019.

Mostrar solo las filas de df_paro que corresponden a un total mayor de 15.

Mostrar solo las filas que corresponden al Periodo 2019 y tienen Total mayor de 15.

## <ins>Ejercicios Opcionales </ins>

## EJERCICIO 18

Investiga el método .sort_values() para ordenar las observaciones según la longitud de las palabras en orden descendente.

## EJERCICIO 19

Convierte la columna vowels a lista y, con sorted() ordénala de mayor a menor. Investiga para ello el método .tolist().

## EJERCICIO 20

Recupera aquellos pisos que ocupan una posición múltiplo de 100:

## EJERCICIO 21

Muestra el piso que se encuentra en la posición 7:

## EJERCICIO 22

Muestra el piso cuyo índice vale 9:

## EJERCICIO 23

Muestra el campo Dirección de los tres primeros pisos. Utiliza para ello tanto el operador ```[]``` como ```[[]]```. ¿Qué diferencia encuentras?

## EJERCICIO 24

Muestra los campos Inmobiliaria y Tipo de inmueble de los tres últimos pisos:

## EJERCICIO 25

Muestra el valor del DataFrame contenido en la posición (2,2):

## EJERCICIO 26

Muestra el valor del DataFrame cuyo índice es 2 y cuya columna es Título:

## EJERCICIO 27

Muestra los tres primeros inmuebles de Tipo de inmueble = Apartamento:

## EJERCICIO 28

Muestra los dos primeros inmuebles de Tipo de inmueble = Piso y que no se alquilen por inmobiliaria (Inmobiliaria = False):

## EJERCICIO 29

Elimina del DataFrame contenido en fotocasa2_df los alquileres de la posición 597 y 598. Asigna el resultado a fotocasa3_df: